app/src/main/java/com/example/kotlinmigration/viewmodels/MainViewModel.kt

```diff
@@ -64,4 +72,97 @@ class MainViewModel: ViewModel() {
             })
         }
     }
-}
\ No newline at end of file
+
+    fun insertDataToDatabase() {
+
+        viewModelScope.launch(Dispatchers.IO) {
+
+
+            _retrofitState.collect() {
+
+                when (it) {
+
+                    is RetrofitEvent.Successful -> {
+                        if (it.response != null) {
+                            for (item in it.response) {
+                                    val convertedData = PostDto(
+                                        Body = item.body,
+                                        ID = item.id,
+                                        UserID = item.userId,
+                                        Title = item.title
+                                    )
+
+                                App.room.postDao().addPost(convertedData)
+                            }
+                        }
+
+                    }
+                    is RetrofitEvent.Failed -> {}
+                    RetrofitEvent.Idle -> {}
+                    RetrofitEvent.Running -> {}
+
+
+                }
+
+
+            }
+
+        }
+
+    }
+
+
+    fun deleteData(){
+
+        viewModelScope.launch(Dispatchers.IO) {
+
+            App.room.postDao().deleteAllData()
+
+        }
+    }
+
+    fun readData(){
+
+        viewModelScope.launch(Dispatchers.IO) {
+
+
+            //checking if PostDto is populated by observing a flow.. had to make readAllData() not a suspend fnc. not sure if correct.
```

Comment:
the best way to do this would be to create a SharedFlow in your viewModel

```
    private val _databaseFlow = MutableSharedFlow<DatabaseEvent>(
        replay = 0,
        extraBufferCapacity = 1,
        onBufferOverflow = BufferOverflow.DROP_OLDEST
    )
    val databaseFlow: SharedFlow<DatabaseEvent> = _flow

    sealed class DatabaseEvent {
        data class SuccessfulRead(val data: List<PostDto>) : DatabaseEvent()
    }
```

then inside your `readData()`

all you have to do is read from Room and emit the read data in your flow, which the Activity can observe
```
val postDtoList = App.room.postDao.readAllData()
_databaseFlow.emit(SuccessfulRead(postDtoList))

```

Reply:
-We created sealed class DatabaseEvent and initialized a data class SuccessfulReady which takes in data of type List<PostDto>. This data class is of type DatabseEvent()
	-We initialized a mutable (changeable) SharedFlow that takes in <DatbaseEvent> *this allows it to use the SuccessfulRead function correct?*
	-We initialized a nonmutable SharedFlow to use freely and observe the flow above that is actually looking at the List<PostDto> data
	-Finally, we used _databaseFlow.emit() to observe the data inside our mutable SharedFlow
		-Please let me know if my understanding is correct above, wanna make sure im implementing this and understanding.
	-Next thing I need to do with this (when i implement new UI components such as toolbar or buttons) is access this flow in main.